### PR TITLE
#1089A Timeline hook

### PR DESCRIPTION
**Pull request by CarlosNZ**
First version of a timeline data wrangler for #1089.

Requires back-end branch from: https://github.com/openmsupply/application-manager-server/pull/647

Have generated timeline events according to the general categories listed in [this document](https://docs.google.com/document/d/1LWCvv42dHwcv4yCWKG26MJI8yZ6QGHAnSwpY7-rWblQ/edit#) (final page). 

I think we can still have some further discussion as to the trade-off between back-end and front-end logic. Obviously there is quite a bit in both currently, with some of the limitations I mentioned in the above PR. We can talk more about this balance if we need too. Personally I think it might be easier to just put everything in "details" (there's enough there already) and let the front-end work it out from there, but it can work either way. We can discuss.

Currently this just loads a raw JSON readout onto the Overview tab -- haven't done any UI work, that'll happen in #1083 , just want to make sure we standardise a format before progressing further.

Have abstracted all the messy logic away into a `useTimeline` hook, so that the Overview page will receive all the data nicely formatted for easy display.

To test, use the Structure mentioned in the back-end PR, start going through the application process, and note the JSON display which appears on the "Overview" tab (review page) as you go through the various steps.

### CODE REVIEW

**Comment by nmadruga** [comment]
Wow - this is going fast track : )

I just finished having an initial look on the Back-end PR and think I'll continue on this tomorrow since I would like to jump back to my other work for a while. Good to see your progress on all that

**Review by nmadruga — COMMENTED** [review]
I think I'm still trying to get all the work that was done here still. Would like to see in more detail at some point. But will need to get back to the other work before trying to help with investigating the issues found in #1093

File: src/utils/hooks/useTimeline/eventMap.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        displayString: () => this.strings.TIMELINE_APPLICATION_REJECTED,
+      },
+    },
+    ASSIGNMENT: {
+      Assigned: {
+        eventType: TimelineEventType.AssignedByAnother,
+        displayString: ({ reviewer, sections, assigner }) =>
+          this.strings.TIMELINE_ASSIGNED.replace('%1', `**${reviewer?.name}**`)
+            .replace('%2', stringifySections(sections))
+            .replace('%3', `**${assigner?.name}**`),
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/utils/hooks/useTimeline/eventMap.ts:78]: Agreed, will do.

**Comment by CarlosNZ** [comment]
Note, have refactored as mentioned in [this comment](https://github.com/openmsupply/application-manager-web-app/pull/1093#issuecomment-1027783530) to simplify the logic etc. I have made a few additional changes to this hook when updating the presentation component (#1093), so best to review this PR for the overall logic / concept, and that one for particular stylistic or UI comments.

File: src/utils/hooks/useTimeline/useTimeline.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+const useTimeline = (structure: FullStructure) => {
+  const { strings } = useLanguageProvider()
+  const [timeline, setTimeline] = useState<Timeline>()
+  const [error, setError] = useState('')
+  const [loading, setLoading] = useState(true)
+
+  const {
+    data,
+    loading: apolloLoading,
```
Comment (nmadruga) [inline src/utils/hooks/useTimeline/useTimeline.tsx:21]: You don't need this one. I think

File: src/utils/hooks/useTimeline/types.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  timestamp: Date
+  events: TimelineEvent[]
+}
+
+interface Timeline {
+  stages: TimelineStage[]
+  //   rawLog: ActivityLog[]
+}
+
+enum TimelineEventType {
```
Comment (nmadruga) [inline src/utils/hooks/useTimeline/types.ts:27]: This is perfect! That's exactly what I was looking for when we chatted ealier. Thank you

File: src/utils/hooks/useTimeline/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  const reviewDecision = possibleDecisions.reduce((closest, e) => {
+    const reviewUnixTimestamp = DateTime.fromISO(event.timestamp).toMillis()
+    return Math.abs(DateTime.fromISO(e.timestamp).toMillis() - reviewUnixTimestamp) <
+      Math.abs(DateTime.fromISO(closest.timestamp).toMillis() - reviewUnixTimestamp)
+      ? e
+      : closest
+  })
+
+  return reviewDecision.details
+}
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/utils/hooks/useTimeline/helpers.ts:27]: Yes, I actually changed this in the subsequent PR. It now gets the closest (in time) decision to the review.

File: src/utils/hooks/useTimeline/useTimeline.tsx (hunk 2 above)
Comment (CarlosNZ) [reply src/utils/hooks/useTimeline/useTimeline.tsx:21]: Yup, sure. 40fd99f0

**Comment by CarlosNZ** [comment]
@nmadruga Can I assume this is fine now too, since I've made most of the changes in the subsquent PR #1093 which is approved?

**Review by nmadruga — APPROVED** [review]

**Pull request merged**